internal/shim/index.js

```diff
@@ -56,44 +92,58 @@ proc.on('exit', function(code, signal){
  * Newline-delimited JSON stdout.
  */
 
-var out = byline(proc.stdout)
-
-out.on('data', function(line){
-  if (debug) console.log('[shim] parsing: `%s`', line)
-
-  var msg;
-  try {
-    msg = JSON.parse(line);
-  } catch (err) {
-    console.log('[shim] unexpected non-json line: `%s`', line);
-    return
+// Chunks holds onto partial chunks received in the absense of a newline.
+// invariant: an array of Buffer objects, all of which do not have any newline characters
+let chunks = [];
+
+// Find successive newlines in this chunk, and pass them along to `handleChunk`
+function handleChunk(chunk) {
+  let startPos = 0;
+  for (;;) {
+    const pos = chunk.indexOf(NEWLINE, startPos);
+
+    // We were not able to find any more newline characters in this chunk,
+    // save the remaineder in `chunks` for later processing
+    if (pos === -1) {
+      chunks.push(chunk.slice(startPos));
+      break;
+    }
+
+    // We have found a whole line
+    const start = chunk.slice(startPos, pos);
+
+    chunks.push(start);
+    const line = Buffer.concat(chunks);
+    chunks = [];
+
+    // set the new start position to one _after_ the last newline seen, so we don't see it again
+    startPos = pos + 1;
```

Comment:
I might be reading this wrong, but should this reset to `0` as well since we're resetting the `chunks` to `[]`?

Reply:
Maybe I'll comment this a bit better, and add a new commit.

`chunks` is an array of buffers that stores all previous data chunks (`Buffer`) that didn't have a newline yet (so you can't turn them into a line)

Since the current chunk can have multiple lines inside of it, `startPos` is an index into the current chunk (which is a Buffer) not `chunks` (which is an array of Buffers).

For a chunk like `['a', 'b', 'c', '\n', 'd', 'e', 'f']`  Start pos will be 0 (aka 'a') then 4 (aka 'd').  The the slice `['d', 'e', 'f']` will be saved in chunks until we receive the next chunk of data.

`chunks = [Buffer ['d', 'e', 'f']];`

Reply:
oops I misread that, thanks for clarifying!